### PR TITLE
Numpy-like API [WIP]

### PR DESCRIPTION
**Pull request by farizrahman4u**
## What works as of now :

* numpy-like indexing and slicing

```python
import jumpy as jp
import numpy as np

x = jp.array(np.random.random((64, 32, 16)))

x.shape  # >> (64, 32, 16)
x[10].shape  # >> (32, 16)
x[:, :10, :].shape  # >> (64, 10, 16)

# Note: nd4j doesn't have 1d vectors but jumpy abstracts this.
x[5][10].shape  # >> (16,)
```

* basic ops

```python
x = jp.zeros((32, 16))
y = jp.ones((32, 16))

z = x + y * 3
```

* broadcasting works (done at python level - will have overhead)

```python
x = jp.zeros((32, 16, 8))
y = jp.zeros((16, 8))

z = x + y
```

### CODE REVIEW

**Comment by farizrahman4u** [comment]
## What works as of now :

* numpy-like index (almost)

```python
import jumpy as jp
import numpy as np

x = jp.array(np.random.random((64, 32, 16)))

x[10].shape  # >> (32, 16)
x.shape  # >> (64, 32, 16)
x[:, :10, :].shape  # >> (64, 10, 16)

# Note: nd4j doesn't have 1d vectors but jumpy abstracts this.
x[5][10].shape  # >> (16,)
```

* basic ops

```python
x = jp.zeros((32, 16))
y = jp.ones((32, 16))

z = x + y * 3
```

* broadcasting works (done at python level - will have overhead)

```python
x = jp.zeros((32, 16, 8))
y = jp.zeros((16, 8))

z = x + y
```

File: jumpy/ndarray.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            self.array = x.add(y)
+        return self
+
+    def __isub__(self, other):
+        other = _indarray(other)
+        if self.array.shape() == other.shape():
+            self.array = self.array.subi(other)
+        else:
+            x, y = broadcast(self.array, other)
+            self.array = x.sub(y)
```
Comment (agibsonccc) [inline jumpy/ndarray.py]: @raver119 what do we want to do here? I know the new ops support auto broadcast now.

File: jumpy/ndarray.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+            return 1
+        return len(self.array.shape())
+
+    @property
+    def ndim(self):
+        return len(self.array.shape())
+
+    def __getitem__(self, key):
+        if self.is0d:
+            raise IndexError('Invalid index to scalar variable.')
```
Comment (agibsonccc) [inline jumpy/ndarray.py]: We should try to have conversions for the other various indices like interval, specified index,.. etc

Comment (farizrahman4u) [reply jumpy/ndarray.py]: Intervals and specified indices do work:
```python
x[0]
x[0, :10]
x[5: 10]
```

**Pull request merged**